### PR TITLE
Test Matrix

### LINKED ISSUE #10658: Expand CI unit test matrix to include windows and macOS
Argues for doing a unit test run on windows and macOS as a matrix really to make sure our dev environments run out of the box, we've started taking some entropy there and used to test them on travis. I'll add an issue

_Originally posted by @mikehardy in https://github.com/ankidroid/Anki-Android/pull/10652#pullrequestreview-926071599_

### PR DESCRIPTION
**Pull request by gpauer**
Updated to run tests with matrix for Windows, macOS, and Linux

## Fixes
Fixes #10658

## Approach
Matrix runs it for Windows, Mac, and Linux.
Timeout had to be increased because when testing windows would be very slow sometimes.
Latest windows uses powershell and not bash so I had to change running 2 commands from "&&" to "|" for multi line commands.

## How Has This Been Tested?
Run unit tests for all 3 systems on GitHub actions.

## Checklist
_Please, go through these checks before submitting the PR._

- ✓ You have not changed whitespace unnecessarily (it makes diffs hard to read)
- ✓ You have a descriptive commit message with a short title (first line, max 50 chars).
- ✓ Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- ✗ You have commented your code, particularly in hard-to-understand areas
- ✓ You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by mikehardy — APPROVED** [review]
This looks great! I can see it running:
https://github.com/gpauer/test_matrix/actions/runs/2158203470
I see it had problems with the coverage upload, I assume it's missing the token in your repo
If it fails to upload coverage post-merge, it should be an easy fix

**Comment by gpauer** [comment]
> This looks great! I can see it running: https://github.com/gpauer/test_matrix/actions/runs/2158203470 I see it had problems with the coverage upload, I assume it's missing the token in your repo If it fails to upload coverage post-merge, it should be an easy fix

Thanks, yep no token.

File: .github/workflows/tests_unit.yml [hunk 1]
```diff
@@ -80,4 +84,6 @@ jobs:
           timeout_minutes: 10
           retry_wait_seconds: 60
           max_attempts: 3
-          command: curl https://codecov.io/bash -o codecov.sh && bash ./codecov.sh 
+          command: |
+            curl https://codecov.io/bash -o codecov.sh
+            bash ./codecov.sh
```
Comment (david-allison) [inline .github/workflows/tests_unit.yml:89]: We might want to use `shell:` here to ensure that Windows uses `git bash` when executing:

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell

File: .github/workflows/tests_unit.yml [hunk 2]
```diff
@@ -31,8 +31,12 @@ on:
 jobs:
   unit:
     name: JUnit Tests
-    timeout-minutes: 20
-    runs-on: ubuntu-latest
+    timeout-minutes: 30
```
Comment (david-allison) [inline .github/workflows/tests_unit.yml:34]: This change doesn't seem documented, but seems reasonable, we're sadly very close to 20 mins on the other platforms

Comment (mikehardy) [reply .github/workflows/tests_unit.yml:34]: gpauer mentioned it in PR description, windows apparently is very slow sometimes (probably the windows defender scanning the files as they unpack etc...)

Comment (david-allison) [reply .github/workflows/tests_unit.yml:34]: D'oh, they did!

File: .github/workflows/tests_unit.yml (hunk 1 above)
Comment (gpauer) [reply .github/workflows/tests_unit.yml:89]: I don't know to be honest, what would be the disadvantages of using powershell?

Comment (gpauer) [reply .github/workflows/tests_unit.yml:89]: but yeah its probably best to use bash for CI, so I will add it

Comment (david-allison) [reply .github/workflows/tests_unit.yml:89]: Not too much, but we can't use `&&` in powershell, and it's good to have a fairly consistent environment

**Review by david-allison — APPROVED** [review]
Kicked off CI, looks good to me, thanks!

**Comment by david-allison** [comment]
All good if CI is green

**Comment by gpauer** [comment]
Whoops

**Comment by david-allison** [comment]
Hmm.. flaky test:

```
com.ichi2.anki.DeckPickerTest > version16CollectionOpens[0] FAILED
    java.lang.AssertionError: Collection should now be open
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:26)
        at com.ichi2.anki.DeckPickerTest.version16CollectionOpens(DeckPickerTest.java:411)
thread '<unnamed>' panicked at 'DBError { info: "SqliteFailure(Error { code: Unknown, extended_code: 1 }, Some(\"no such table: decks\"))", kind: Other }', src/lib.rs:256:13
```

**Comment by gpauer** [comment]
Revert to PowerShell for now? Then perhaps we can create issue to fix flaky test and put bash back

**Comment by david-allison** [comment]
I've seen this flake a few times on my Windows box - might be best to ignore the test for now

**Comment by gpauer** [comment]
> I've seen this flake a few times on my Windows box, it's not a PowerShell issue - might be best to ignore the test for now

So should I close the issue or we keeping it open for now?

**Review by mikehardy — APPROVED** [review]

**Comment by mikehardy** [comment]
Hi there! APRIL OpenCollective Notice

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for [April 2022 submissions](https://github.com/ankidroid/Anki-Android/pulls?q=is%3Apr+merged%3A2022-04-01..2022-04-30)

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged for in the month of March.

Please note that GSoC contributions are okay for this process. Our philosophy is that our users have donated to AnkiDroid for all contributions. The only PRs that will not go through the OpenCollective process are ones directly related to am accepted GSoC project from a selected participant, since those receive a stipend from GSoC itself.

Please understand that this means the monthly budget we have for payments might be stretched a bit and the cap on payments-per-person may be lower, but we try to make our process as fair and transparent as possible, we just need your understanding.

Thanks!

**Pull request merged**